docs/sdg/sdg-flow-yaml.md

```diff
@@ -0,0 +1,101 @@
+# SDG API :: Add a file format for defining custom Flows
+
+## Problem Statement
+
+The `instructlab/sdg` library is introducing more extensive data generation pipelines. To enable customization, we should allow users of the library to provide a configuration file which defines a custom pipeline or extends an existing pipeline.
+
+In terms of the API constructs, a Pipeline is created from a sequence of “block configurations” which express how to instantiate and invoke the individual steps (aka blocks) in the pipeline. A Flow construct serves as a template from which a sequence of block configs can be generated.
+
+## Objective
+
+- Library users can specify a custom flow using a well-defined file format.
+- Library users can either use a custom flow standalone, or combine a custom flow with existing flows.
+- The file format and library can evolve substantially without making breaking changes.
+- Incompatible changes can be introduced while retaining support for existing custom flows for a deprecation period.
+
+## Proposal
+
+We will add an API that instantiates a `Flow` object from a file:
+
+```python
+flow = Flow.from_file('mycustomflow.yaml', params)
+block_configs = flow.render()
+```
+
+The YAML file format will mirror the API and look like this:
```

Comment:
I wonder if we should put a version field on this in case we need to break compatibility in the future?

https://github.com/instructlab/sdg/issues/54

I hope it would also serve as a reminder to anyone poking around in the code that the format is not arbitrary and changes must consider compatibility.

Reply:
Yes! I even put that in the objectives and then did nothing about it!

Reply:
and I should have read the whole thing properly before commenting :)

Reply:
Added in commit 1781b8a thanks